File: Dynamic programming/leetcode_73.py

# Problem link ===>> https://leetcode.com/problems/set-matrix-zeroes/?envType=problem-list-v2&envId=21r4yhih
# submission link == =>> https://leetcode.com/problems/set-matrix-zeroes/submissions/1787334098/?envType=problem-list-v2&envId=21r4yhih

import sys
import math
import bisect
import heapq
from collections import defaultdict, deque, OrderedDict, Counter
from functools import lru_cache, reduce
from itertools import permutations, combinations, accumulate
from typing import List, Tuple, Dict, Set, Optional, Union, Any

MOD = 10**9 + 7
INF = float('inf')
input = sys.stdin.read
print = sys.stdout.write

class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        zeroinFirstCol = False
        for row in range(len(matrix)):
            if matrix[row][0] == 0:
                zeroinFirstCol = True
            for col in range(1, len(matrix[0])):
                if matrix[row][col] == 0:
                    matrix[row][0] = 0
                    matrix[0][col] = 0

        for row in range(len(matrix) - 1, -1, -1):
            for col in range(len(matrix[0]) - 1, 0, -1):
                if matrix[row][0] == 0 or matrix[0][col] == 0:
                    matrix[row][col] = 0
            if zeroinFirstCol:
                matrix[row][0] = 0

def main():
    t = int(input())
    for _ in range(t):
        n = int(input())

if __name__ == "__main__":
    solution = Solution()

File: Dynamic programming/leetcode_73.cpp

// Problem link ===>> https://leetcode.com/problems/set-matrix-zeroes/?envType=problem-list-v2&envId=21r4yhih

#include <bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp>
#include <ext/pb_ds/tree_policy.hpp>
using namespace std;
using namespace __gnu_pbds;
#define ll long long
#define ld long double
#define endl '\n'
const int MOD = 1e9 + 7;
const double EPS = 1e-9;
const double PI = acos(-1.0);
template <typename T>
using ordered_set = tree<T, null_type, less<T>, rb_tree_tag, tree_order_statistics_node_update>;
template <typename K, typename V>
using ordered_map = tree<
    K, V, less<K>, rb_tree_tag, tree_order_statistics_node_update>;

class Solution
{
public:
    void setZeroes(vector<vector<int>> &matrix)
    {
        bool zeroFirstCol = false;

        for (int row = 0; row < matrix.size(); row++){
            if(matrix[row][0] ==0)
                zeroFirstCol = true;
            for (int col = 1; col < matrix[0].size(); col++){
                if(matrix[row][col] == 0){
                    matrix[row][0] = 0;
                    matrix[0][col] = 0;
                }
            }
        }

        for (int row = matrix.size() - 1; row >= 0; row--){
            for (int col = matrix[0].size() - 1; col >= 1; col--){
                if(matrix[row][0] == 0 || matrix[0][col] == 0){
                    matrix[row][col] = 0;
                }
            }
            if(zeroFirstCol){
                matrix[row][0] = 0;
            }
        }
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cout.tie(nullptr);
    cin.tie(nullptr);
    
    ll t;
    cin >> t;
    while (t--)
    {
        ll n;
        cin >> n;

        
    }
    return 0;
}